### ISSUE TITLE
`uv tool install` unable to install a tool that depends on intermediate dependencies with entrypoints

### ISSUE THREAD

**Issue by j178** [astral-sh/uv#4994]
For instance, `fastapi` doesn't have any entrypoints but it depends on `fastapi-cli`, which provides an entrypoint named `fastapi`. 
`uv tool install fastapi` would result in a failure, displaying the error message `No executables found for fastapi`.

```
$ uv tool install fastapi
warning: `uv tool run` is experimental and may change without warning.
Resolved 36 packages in 729ms
Installed 36 packages in 92ms
...
 + fastapi==0.111.0
 + fastapi-cli==0.0.4
...
error: No executables found for `fastapi`
```

But at the same time `uv tool run fastapi` works:
```
$ uv tool run fastapi --version
warning: `uv tool run` is experimental and may change without warning.
Resolved 36 packages in 18ms
FastAPI CLI version: 0.0.4
```

## Environment
```
os: Windows 11 Pro 23H2 (Build 22631.3737)
uv version: 0.2.24
```

**Comment by zanieb** [comment]
Interesting. I don't think we should install executables from dependencies by default, but we should allow opt-in.

**Comment by j178** [comment]
`rye tools install` has a flag `--include-dep <INCLUDE_DEP>` to include scripts from a given dependency, maybe `uv` can allow something likes this.

**Comment by zanieb** [comment]
I actually won't take this on right away, I've got to take care of some other things first if someone is interested. I think we need something like:

- `--executable <name>`: Only install the given executable name
- `--include-package <package name>`: Include executables from the given dependency package

I'm a little unsure of the name for the second one. Maybe:

- `--include-dep`
- `--include-executables`
- `--executables-from`

For `--executable`, we need to decide if we'll scan dependencies if it's not found (we probably shouldn't for safety and performance but we _could_). If not, it's kind of separate from this issue.

We'll need to add included packages to the receipt, which is where most of the complexity is here.

**Comment by charliermarsh** [comment]
Why would this not be uv tool install fastapi-cli?

**Comment by charliermarsh** [comment]
And if `fastapi-cli` doesn't depend on `fastapi`, couldn't you do: `uv tool install fastapi-cli --with fastapi`?

**Comment by charliermarsh** [comment]
Although it seems like `uvx fastapi --from fastapi-cli` just works?

**Comment by zanieb** [comment]
Yeah `uvx --from fastapi-cli fastapi ...` does seem like the proper invocation here — but we would need to support including dependency entry points for parity with Rye and pipx.

**Comment by charliermarsh** [comment]
Yeah it just seems like it's kind of... the wrong way around, to include a dependency entrypoint. And it complicates the CLI a lot.

**Comment by zanieb** [comment]
I agree it's a bit awkward. I think we should pursue #5017 and #5018 to help guide users to the correct path here and we can consider implementing this more complicated API based on additional feedback.

**Comment by charliermarsh** [comment]
Ok, `uv tool run jupyter` is a counterexample here. Right now, we suggest `uv tool run --from jupyter-core jupyter`. And it's true that `jupyter-core` contains the Jupyer executable... but you _also_ need `jupyer` installed in order to do, like, `jupyter notebook`. Otherwise, you get: `Jupyter command `jupyter-notebook` not found.`

So the command in that case would really need to be... `uv tool run --from jupyter-core --with jupyter jupyter notebook`.

**Comment by dmvianna** [comment]
This is something that needs to be more salient in the documentation. Many **Python** executables depend or are modified by optional dependencies. `jupyter` is an example, `dagster` is another. I wasn't able to run the development server until I saw the comment above.

So, from 

```sh
$ pip install dagster dagster-webserver
$ dagster dev
```

I was able to execute it straight from the `./venv` with

```sh
$ uv add dagster dagster-webserver
$ uv tool run --from dagster-webserver dagster dev
```

Not as short, but it works! 💯
